### DISCUSSION TITLE
company-conf error: Can not find python

### DISCUSSION THREAD

**Discussion by buhtz** [company-mode/company-mode#1308] [Q&A] [answered]
I want to use `company` in combination with `lsp-mode` and `python-mode` [as described by SystemCrafters (aka David Wilsion)](https://github.com/daviwil/emacs-from-scratch/blob/dd9320769f3041ac1edca139496f14abe147d010/Emacs.org#python).

But I got this error after `M-x company-mode`:

`Company: backend company-capf error "Searching for program: File or Folder not found, python" with args \ (prefix)`

It IMHO indicates that it can not find the python interpreter `python`, which really does not exist but `python3` does and is set up in the `python-mode`.

This is an "minimal" `init.el` to reproduce the problem.

```lisp
;; -*- lexical-binding: t; -*-

;; === Package setup ===
(require 'package)

(setq package-archives '(("melpa" . "https://melpa.org/packages/")
			 ("org" . "https://orgmode.org/elpa/")
			 ("elpa" . "https://elpa.gnu.org/packages/")))
;; Initializes the package infrastructure
(package-initialize)

;; === use-package ==
;; use-package to simplify the config file
(unless (package-installed-p 'use-package)
  (package-refresh-contents)
  (package-install 'use-package))

(require 'use-package)
(setq use-package-always-ensure 't)

;; === Improved search field
; M-o opens "context menu" of a search item
(use-package ivy
  :diminish
  :bind (("C-s" . swiper))
  :config
  (ivy-mode 1)
  (setq ivy-use-selectable-prompt t)
  )

;; === COUNSEL
; add some extra infos to some minibuffers
; e.g. ivy-rich need it
(use-package counsel
  :bind (("M-x" . counsel-M-x)
	 ("C-x b" . counsel-switch-buffer)  ; counsel-ibuffer
	 ("C-x C-f" . counsel-find-file)))

;; === Add docstring column to function search results
(use-package ivy-rich
  :init
  (ivy-rich-mode 1))

;; === EVIL
(use-package evil
  :init
  (setq evil-want-integration t)
  (setq evil-want-keybinding nil)
  (setq evil-want-C-i-jump nil)
  :config
  (evil-mode 1)
  ;; Use visual line motions even outside of visual-line-mode buffers
  (evil-global-set-key 'motion "j" 'evil-next-visual-line)
  (evil-global-set-key 'motion "k" 'evil-previous-visual-line)
  ;; Set "normal" vi-mode in specific buffers
  (evil-set-initial-state 'messages-buffer-mode 'normal)
  (evil-set-initial-state 'dashboard-mode 'normal)
  )

(use-package evil-collection
  :after evil
  :config
  (evil-collection-init)
  )

; lsp-find-definiton
; lsp-find-reference
; lsp-rename
(use-package lsp-mode
  :commands (lsp lsp-deferred)
  :init
  (setq lsp-keymap-prefix "C-l")
  :config
  (lsp-enable-which-key-integration t)
  )

(use-package python-mode
  :ensure t
  :hook
  (python-mode . lsp-deferred)
  :custom
  (python-shell-interpreter "python3")
  )

; pip3 install 'python-language-server[all] 
(use-package company
  :after lsp-mode
  :hook (lsp-mode . company-mode)
  :bind (:map company-active-map
         ("<tab>" . company-complete-selection))
        (:map lsp-mode-map
         ("<tab>" . company-indent-or-complete-common))
  :custom
  (company-minimum-prefix-length 1)
  (company-idle-delay 0.0))

(use-package company-box
  :hook (company-mode . company-box-mode))


```

I assume that ivy, counsel and evil can be removed from that init.el. It was just there for my comfort.

**Comment by dgutov** [comment]
Hi!

Try producing the error backtrace by `M-x toggle-debug-on-error` and then repeating the action that leads to the error message.

**Comment by buhtz** [comment]
It is obvious that it looks for `python`.

But...
1. `python` is not available/installed on that system but `python3` is.
2. I still setup `python3` in the `python-mode` config (see `python-shell-interpreter "python3"`).

```lisp
Debugger entered--Lisp error: (file-missing "Searching for program" "Datei oder Verzeichnis nicht gefunden" "python")
  make-process(:name "python" :buffer "*Python Fast*" :command ("python" "-u" "-i"))
  apply(make-process (:name "python" :buffer "*Python Fast*" :command ("python" "-u" "-i")))
  start-process("python" "*Python Fast*" "python" "-u" "-i")
  apply(start-process "python" "*Python Fast*" "python" ("-u" "-i"))
  py-shell(nil nil nil "python" nil t)
  py--fast-complete-base(nil "g" "import json;import os;import os.path;import os.j;i...")
  py-shell-complete(nil nil nil nil 1 nil)
  py-fast-complete()
  completion--capf-wrapper(py-fast-complete optimist)
  run-hook-wrapped(completion--capf-wrapper py-fast-complete optimist)
  company--capf-data-real()
  company--capf-data()
  company-capf(prefix)
  apply(company-capf prefix)
  company-call-backend-raw(prefix)
  apply(company-call-backend-raw prefix)
  company--force-sync(company-call-backend-raw (prefix) company-capf)
  #f(compiled-function (&rest args) #<bytecode 0x2aaab58f4cfd>)(prefix)
  apply(#f(compiled-function (&rest args) #<bytecode 0x2aaab58f4cfd>) prefix)
  company-call-backend(prefix)
  company--begin-new()
  company--perform()
  company-auto-begin()
  company-idle-begin(#<buffer x.py> #<window 1 on x.py> 90 823)
  apply(company-idle-begin (#<buffer x.py> #<window 1 on x.py> 90 823))
  timer-event-handler([t 25098 4852 240216 nil company-idle-begin (#<buffer x.py> #<window 1 on x.py> 90 823) nil 761000])
```

I am totally new to Lisp. Maybe this code point is relevant?
https://github.com/company-mode/company-mode/blob/c25f1fbc3850e36e6521b77fa1641d5583365d8b/company-capf.el#L73

**Reply by dgutov** [reply]
What is that `py-fast-complete` function? Which package does it come from?

**Reply by buhtz** [reply]
`C-h f` tells me
```lisp
py-fast-complete is an interactive compiled Lisp function in
‘python-mode.el’.

(py-fast-complete &optional SHELL WORD IMPORTS)

Complete word before point, if any.

Use ‘py-fast-process’
```
It is in `python-mode.el` (Version 6.3.1).

I can not finde `python-mode` in the `M-x package-install` seach menu. But the `python-mode.el` is stored in the directory `~/.emacs.d/elpa/python-mode-20211229.1002/`. This confuses me because I thougth that this mode is Emacs in-build. I am using Emacs 27 from the Debian stable (11, Bullseye) repository.

**EDIT**:
I did a full update of all installed packages via `package-list` + `U` + `x`. The behavior did not change. Also the version of `python-mode` did not change.

**Reply by dgutov** [reply]
Apparently there is a package called `python`. It's built-in, and also distributed through GNU ELPA: https://elpa.gnu.org/packages/python.html

And there is *different one* called `python-mode` in MELPA: https://melpa.org/#/python-mode, with upstream at https://gitlab.com/python-mode-devs/python-mode/

**Reply by dgutov** [reply]
You can try uninstalling it. Or double down and find which custom var that package uses, to change `python` to `python3` there.

**Reply by buhtz** [reply]
I am not sure if I understand. The `python-mode` does not appear in the `M-x list-packages` list. Why?

Anyway. Can you give an config example about how to setup your company mode with the usual inbuild (ELPA) python mode?

**Reply by dgutov** [reply]
> I am not sure if I understand. The python-mode does not appear in the M-x list-packages list. Why?

Perhaps because it's already installed? It's should be listed as "Installed", though.

> Can you give an config example about how to setup your company mode with the usual inbuild (ELPA) python mode?

No special config, you call `M-x run-python` and then enjoy some completion that is based on the Python shell's introspection. I don't really work with this language and thus have little experience with this setup. I'm not sure if you need to `import` packages in the shell buffers to see completions from them, or not.

Finally, if you want feature-rich completion (with smarted type analysis, docstrings, icons, etc), your best bet is likely to install `lsp-mode` or `eglot` and the corresponding Language Server. IIUC for Python the current recommendation is `pyright`.

**Comment by andreas-roehler** [comment]
AFAUI you detected correctly, that it can not find the interpreter python. But that's a system issue, not an Emacs thing. Something like` ln -s python3 python` may help.

**Comment by andreas-roehler** [comment]
WRT python-mode.el: if a executable `python` doesn't exist, the mode should deal with it. Either give a warning or select python3.  Created a ticket for this: [https://gitlab.com/python-mode-devs/python-mode/-/issues/117]( https://gitlab.com/python-mode-devs/python-mode/-/issues/117)

**Comment by buhtz** [comment, marked as answer]
The hint about that two different and overriding python modes was very helpful. Now I could setup the Emacs 27 in-build python mode with company mode. The result `init.el` can be found here: https://github.com/company-mode/company-mode/discussions/1309
